### PR TITLE
Add ability to scale on tgi custom metrics

### PR DESCRIPTION
**Pull request by rsgowman**

### CODE REVIEW

File: benchmarks/inference-server/text-generation-inference/custom-metrics-stackdriver-adapter/main.tf [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+  metadata {
+    name = "custom-metrics-stackdriver-adapter"
+    namespace = kubernetes_namespace_v1.custom-metrics.metadata[0].name
+    annotations = {
+      "iam.gke.io/gcp-service-account" = google_service_account.cmsa-sa[0].email
+    }
+  }
+}
+
+# This doesn't work. It fights with the service account for control over the
```
Comment (annapendleton) [inline benchmarks/inference-server/text-generation-inference/custom-metrics-stackdriver-adapter/main.tf:26]: Should we remove this commented out section?

File: benchmarks/inference-server/text-generation-inference/custom-metrics-stackdriver-adapter/README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# Custom Metrics Stackdriver Adapter
```
Comment (annapendleton) [inline benchmarks/inference-server/text-generation-inference/custom-metrics-stackdriver-adapter/README.md:1]: No action here, just an interesting possibility for potential work for the custom metrics stackdriver adapter modules

We've been utilizing infrastructure modules from Cloud Foundry (eg. example for GKE Hub, https://github.com/GoogleCloudPlatform/cloud-foundation-fabric/tree/master/modules/gke-hub) when possible for getting default GKE configurations and easier TF automation.

Wonder if adding some modules there for GKE HPA at some point would also be a great place for customers to find these generic modules for configuration.

File: benchmarks/inference-server/text-generation-inference/sample-terraform.tfvars [hunk 3]
```diff
@@ -9,13 +9,19 @@ gpu_count = 1
...
 # - Other possibilities coming soon...
 #
-# Note: "cpu" scaling is a poor choice for this workload - the tgi workload
-# starts up, pulls the model weights, and then spends a minute or two worth of
-# cpu time crunching some numbers. This causes hpa to add a replica, which then
-# spends more cpu time, which causes hpa to add a replica, etc. Eventually,
-# things settle, and hpa scales down the replicas. This whole process could
-# take up to an hour.
+# See `autoscaling.md` for more details and recommendations.
 hpa_type = null
```
Comment (annapendleton) [inline benchmarks/inference-server/text-generation-inference/sample-terraform.tfvars:21]: Default null case throws an error:

kubernetes_manifest.default["./manifest-templates/text-generation-inference-svc.tftpl"]: Refreshing state...
│ Error: Invalid function argument
│ 
│   on main.tf line 27, in locals:
│   25:     contains(
│   26:       ["tgi_queue_size", "tgi_batch_current_size", "tgi_batch_current_max_tokens"],
│   27:       var.hpa_type
│   28:     )
│     ├────────────────
│     │ while calling contains(list, value)
│     │ var.hpa_type is null
│ 
│ Invalid value for "value" parameter: argument must not be null.
╵

File: benchmarks/inference-server/text-generation-inference/main.tf [hunk 4]
```diff
@@ -18,6 +18,16 @@ locals {
 
   all_templates = concat(local.wl_templates, local.secret_templates)
 
+  hpa_cpu_template = "${path.module}/hpa-templates/hpa.cpu.yaml.tftpl"
+  hpa_custom_metric_template = "${path.module}/hpa-templates/hpa.tgi.custom_metric.yaml.tftpl"
+  tgi_podmonitoring = "${path.module}/hpa-templates/tgi-podmonitoring.yaml.tftpl"
+  custom_metrics_enabled = (
+    contains(
+      ["tgi_queue_size", "tgi_batch_current_size", "tgi_batch_current_max_tokens"],
```
Comment (laoj2) [inline benchmarks/inference-server/text-generation-inference/main.tf:26]: Can we set custom_metrics_enabled to false only when hpa_type is `null` or `cpu`, instead?

File: benchmarks/inference-server/text-generation-inference/sample-terraform.tfvars [hunk 5]
```diff
@@ -9,13 +9,19 @@ gpu_count = 1
...
+
+# Sets the averagevalue target of the hpa metric. e.g for cpu scaling, this is
+# the cpu utilization, expressed as a value between 0-100.
+hpa_averagevalue_target = 123  # TODO: Experiment with this to determine optimal values
```
[the remark this answers is not in the document]
Comment (annapendleton) [reply benchmarks/inference-server/text-generation-inference/sample-terraform.tfvars]: lgtm, defs a reasonable place to start :)

**Review by annapendleton — APPROVED** [review]

**Pull request merged**